eth/blockwatch/block_watcher.go

```diff
@@ -281,28 +284,45 @@ func (w *Watcher) addLogs(header *MiniHeader) (*MiniHeader, error) {
 func (w *Watcher) getMissedEventsToBackfill(ctx context.Context) ([]*Event, error) {
 	events := []*Event{}
 
+	var (
+		startBlockNum          int
+		latestRetainedBlockNum int
+		blocksElapsed          int
+	)
+
 	latestRetainedBlock, err := w.stack.Peek()
 	if err != nil {
 		return events, err
 	}
-	// No blocks stored, nowhere to backfill to
-	if latestRetainedBlock == nil {
-		return events, nil
-	}
+
 	latestBlock, err := w.client.HeaderByNumber(nil)
 	if err != nil {
 		return events, err
 	}
-	blocksElapsed := big.NewInt(0).Sub(latestBlock.Number, latestRetainedBlock.Number)
-	if blocksElapsed.Int64() == 0 {
+	latestBlockNum := int(latestBlock.Number.Int64())
+
+	if latestRetainedBlock != nil {
+		latestRetainedBlockNum = int(latestRetainedBlock.Number.Int64())
+		// Events for latestRetainedBlock already processed, start at latestRetainedBlock + 1
+		startBlockNum = latestRetainedBlockNum + 1
+		blocksElapsed = latestBlockNum - startBlockNum
+	} else if w.backfillStartBlock != nil {
+		startBlockNum = int(w.backfillStartBlock.Int64())
+		blocksElapsed = latestBlockNum - startBlockNum
+	} else {
 		return events, nil
 	}
 
-	glog.Infof("Some blocks have elapsed since last boot. Backfilling block events (this can take a while)... blocksElapsed=%v", blocksElapsed.Int64())
-	startBlockNum := int(latestRetainedBlock.Number.Int64() + 1)
-	endBlockNum := int(latestRetainedBlock.Number.Int64() + blocksElapsed.Int64())
+	if blocksElapsed == 0 {
+		return events, nil
+	}
+
+	glog.Infof("Some blocks have elapsed since last boot. Backfilling block events (this can take a while)...\n")
+	glog.Infof("Start block: %v		 End block: %v		Blocks elapsed: %v\n", startBlockNum, startBlockNum+blocksElapsed, blocksElapsed)
+
+	endBlockNum := startBlockNum + blocksElapsed
 	logs, furthestBlockProcessed := w.getLogsInBlockRange(ctx, startBlockNum, endBlockNum)
-	if int64(furthestBlockProcessed) > latestRetainedBlock.Number.Int64() {
+	if furthestBlockProcessed > latestRetainedBlockNum {
```

Comment:
I don't think we need to execute the logic in this if statement in the scenario where there is no last retained block and we're backfilling starting from `backfillStartBlock` since there would be no blocks in the DB yet.

Reply:
We do need to store the block we backfilled to (last seen block now) in the DB so that when we execute `Watch()` we'll start from that block

So rather than shifting the logic and conditionals around I thought it would be okay to have this statement always execute when there are no blocks in the DB (`lastRetainedBlock = 0`). 

The only operation that is kind of unnecessary is that we fetch the headers from the DB and check the length.

Reply:
Any additional thoughts or are you okay with keeping the functionality as is here?

Reply:
I think this should be ok